### ISSUE TITLE
Other XMS Manager, HMA remains empty

### ISSUE THREAD

**Issue by JayQ2K** [joncampbell123/dosbox-x#2299]
When using another XMS Manager (xms=false) like HIMEM.SYS, HIMEMX.EXE or JEMMEX.EXE, there seems to be no way to get dos=high,umb working anymore.
Feedback continuously is "HMA is free", while with xms=true the message becomes "HMA is allocated" and "MS-DOS is resident in the high memory area."

Config:
install     = MOUNT.COM C: D:\DOSBox\C -freesize 16
device      = C:\_LOAD\HIMEMX\HIMEMX.EXE
devicehigh  = C:\_LOAD\JEMM579B\JEMM386.EXE X=TEST
break       = off
dos         = high, umb
fcbs        = 100
files       = 200
lastdrive   = a
set path    = Z:\
set prompt  = $P$G

Tried moving dos=high,umb upwards, directly below the XMM and before the EMM, but this does not help either.
It seems dos=high is in the order being executed before the rest of [config] and if there is no XMS available at that time DOS will not move into HMA anymore if you run an XMM later on in the process.

**Comment by Wengier** [comment]
I need to mention that currently device driver support is still limited. Hope issues like this will be resolved in the future,

**Comment by JayQ2K** [comment]
For the internal XMM you might want to check the code against HimemX.
In the end I would like to stop using external device drivers alltogether, but for that I do need an environment that sort of exposes all its XMM/EMM capabilities like HIMEM/EMM386 do.

Think to get that going in the right direction it's first to get the XMM part compatible with HimemX. If that's the case, any EMM that might run on top of that should be able to do so without any warnings.
By then you might want to take a stab at the EMM part. My guess there is though that when Internal XMM can expose its capabilities like HimemX does it should be possible to do a full VCPI activation.

As far as I can see few applications truly want EMS, but are more interested in whether VCPI is available (when using MS-DOS EMM386.exe I usually use NOEMS and all my applications work fine...without VCPI most things halt immediately though).

JEMM386 might not be the ideal candidate for DOSBox-X EMM, since it's not fully compatible with MS-DOS EMM386. Maybe FreeDOS 1.0 EMM386 is a better option in that sense.

**Comment by Wengier** [comment]
Sure, it is certainly desired to make the internal XMM implementation more complete and compatible. Meanwhile, you can boot into guest DOS systems in DOSBox-X which can provide full XMM/EMM capabilities with the use of HIMEM/EMM386. You can find more information about booting guest systems in the DOSBox-X Wiki.

Both JEMM386 and FreeDOS EMM386 have their pros and cons, but neither is fully compatible with MS-DOS EMM386. For example, neither support the GEMMIS protocol which is useful for the Windows 3.x support. Their maintainers do not seem to be interested in implementing this even though there are several requests. DOSBox-X (like DOSBox) does implement the GEMMIS protocol on the other hand.

**Comment by JayQ2K** [comment]
I've seen the thread on it at the JEMM GitHub. And that btw definitely isn't the only thing not in line with EMM386. So that probably requires to develop the XMS/ EMS within DOSBox-X completely?
Could you if that is the case first take a look at the HIMEM bit. I think himem.sys and himemx.exe are quite alike and if the functions from that get exposed properly through DOSBox, the internal EMM might also immediately be able to do full dynamic memory allocation.
With that this whole thing could be dropped since by then you can simply use the internal XMM and EMM and have HMA and UMB active at launch.

**Comment by Wengier** [comment]
Sure, we would hope to see complete XMS/EMS handlers in DOSBox-X, perhaps even supporting the new XMS 3.5 API to overcome the 4 GB barrier. But it will apparently require some work. It will be great if you (or anyone else) can help with this too.

**Comment by JayQ2K** [comment]
Took a look through the code and searched for xms, to find "master/src/ints/xms.cpp". It looks like 4309h is not getting checked at all, making the handle table be more conform XMS version 2.0 which had a fixed way to create the handle table.
http://cd.textfiles.com/1stcanadian/utils/xms_spec/ > click on himem.asm (MS HIMEM 2.0 assembly, which Microsoft open-sourced) shows similar behavior.

Looking at HimemX/blob/master/HimemX.asm (from Japheth) there's this, starting at line 586:
@@maybe_my2f:
	cmp al,00h			; is it "Installation Check"?
	je @@get_driver_installed
	cmp al,10h			; is it "Get Driver Address"?
	je @@get_xms_address	 
	cmp al,09h			; is it "get handle table"?
	je @@get_xms_handle_table
	cmp al,08h			; is it "get A20 handler number"?
	jne @@jmp_old2f
	mov al,ah		;al==43h if function supported

Specifically the 09h (get handle table) is what I'm talking about.
Can't say I'm an expert programmer, but reading through the code I can't find any reference to 4309h in the DOSBox-X files.

**Comment by Wengier** [comment]
@JayQ2K You are right that Int2F/AX=4309h is not yet implemented in the code. Please take a look at the DOS_MultiplexFunctions() function of the src/dos/dos_misc.cpp file, as the said function is supposed to be implemented here (along with other cases).

**Comment by JayQ2K** [comment]
So far it seems there is only one thing missing at "DOS_MultiplexFunctions() function", is the 4309 case. This however leads to the addition of multiple lines of additional code to the "master/src/ints/xms.cpp" file. The main thing to it is that this specific case requires code to ensure an XMS handle table gets created so dependent programs can actually show the correct data.

Why exactly should this be in the "src/dos/dos_misc.cpp" file? I don't find any reference to XMS in this file.

**Comment by Wengier** [comment]
@JayQ2K The DOS_MultiplexFunctions function in the src/dos/dos_misc.cpp file is the Int2F handler. It may either handle the cases entirely on its own, or maybe call a function defined in another file (like src/ints/xms.cpp). Hope this helps.

**Comment by JayQ2K** [comment]
I see what you mean and also identified the logic to it. The cases in question (0x4300 and 0x4310) are however already defined in src/ints/xms.cpp (line 363 and 366 respectively). This has for 0x4310 to do with xms_callback that needs some extensive coding. 0x4309 also requires extensive coding to make the case functional. So I would figure it to be ideal in the src/ints/xms.cpp file, since 0x4309 should be part of "static bool multiplex_xms(void)" on line 361 in that file anyway.

**Comment by Wengier** [comment]
@JayQ2K Yes, you are right that you can also add the 0x4309 case to the multiplex_xms function which is a subset of the Int2F handler specific to the XMS.

**Comment by JayQ2K** [comment]
[jemmex.log](https://github.com/joncampbell123/dosbox-x/files/10099499/jemmex.log)
Continuing a bit on this one: When I use a different XMS manager than the internal one, I get a bunch of ERROR CPU messages at the end of the debug log (see attached). Also cpucycles goes up when running Dosbox like that.
